rust/core-lib/src/file.rs

```diff
@@ -250,32 +250,62 @@ where P: AsRef<Path>
         .ok()
 }
 
-impl From<io::Error> for FileError {
-    fn from(src: io::Error) -> FileError {
-        FileError::Io(src)
-    }
-}
-
-impl From<str::Utf8Error> for FileError {
-    fn from(_: str::Utf8Error) -> FileError {
-        FileError::UnknownEncoding
-    }
-}
-
 impl From<FileError> for RemoteError {
-    fn from(_src: FileError) -> RemoteError {
+    fn from(src: FileError) -> RemoteError {
         //TODO: when we migrate to using the failure crate for error handling,
         // this should return a better message
-        RemoteError::custom(5, "failed to load file", None)
+		use std::error::Error;
+		let mut code = 0;
+		let mut message = String::new();
+		match src {
+			FileError::Io(e, p) => {
+				code = 5;
+				match p.to_str() {
```

Comment:
Instead of doing all this string formatting up here, we can do it in the `Display` implementation below, and then up here all we need to do is return `e.to_string()` (`to_string` comes from the `ToString` trait, which you get for free if you implement `Display`).

We could take this a bit further and implement an `error_code()` fn on `FileError`, that just does a match and returns a code; then we could remove this whole match statement, and this whole fn becomes:

```rust
let code = src.error_code();
let message = src.to_string();
RemoteError::custom(code, message, None)
```